### DISCUSSION TITLE
passing a filename (containing user_data commands) as a parameter keeps failing

### DISCUSSION THREAD

**Discussion by Analect** [aws/aws-cdk#30376] [Q&A]
Can anyone point me in the right direction here.
I would like to be able to run up an ec2 instance with differing user_commands (additional software installed) based on different files fed as parameters.

For instance, at the path `'/Users/xxx/dev/gitlab/infrastructure/aws/cdk/cdk-ec2-parameterized/cdk_ec2/user_data/user_data_hello_world.sh'` I have one such file. I am running `cdk synth` from ``'/Users/xxx/dev/gitlab/infrastructure/aws/cdk/cdk-ec2-parameterized/`.

From `cdk_ec2_stack.py`:
```
...
file_name_param = CfnParameter(self, "userDataFileName",
                                       type="String",
                                       description="Name of the file containing user data commands",
                                       default="user_data_hello_world.sh",
                                       ).value_as_string

      
current_dir = os.path.dirname(__file__)
    with open(os.path.join(current_dir, "user_data", file_name_param)) as f:
        ec2_user_data = f.read()
```
Running the following is failing.
```
cdk synth \
--parameters userDataFileName=user_data_hello_world.sh
```

`FileNotFoundError: [Errno 2] No such file or directory: '/Users/xxx/dev/gitlab/infrastructure/aws/cdk/cdk-ec2-parameterized/cdk_ec2/user_data/${Token[TOKEN.19]}'`

However, if I explicitly use `with open(os.path.join(current_dir, "user_data", "user_data_hello_world.sh")) as f` rather than `with open(os.path.join(current_dir, "user_data", file_name_param)) as f`, then it works. 

```
...
file_name_param = CfnParameter(self, "userDataFileName",
                                       type="String",
                                       description="Name of the file containing user data commands",
                                       default="user_data_hello_world.sh",
                                       ).value_as_string

      
current_dir = os.path.dirname(__file__)
    with open(os.path.join(current_dir, "user_data", "user_data_hello_world.sh")) as f:
        ec2_user_data = f.read()
```
What am I missing?

**Comment by Flupinochan** [comment]
Parameters in CloudFormation are believed to be used during CloudFormation deployment. In the code, it seems that parameters of CloudFormation are not being referenced.

Please refer to the following manual for guidance.
TypeScript: https://cdkworkshop.com/20-typescript/40-hit-counter/300-resources.html
Python: https://cdkworkshop.com/30-python/40-hit-counter/300-resources.html

**Reply by Analect** [reply]
@Flupinochan thank you for your input, but I'm not sure those links are relevant to my needs. 

I ended up using a combination of `--context` values from the command-line (supplying the file names) combined with `aws_s3_assets` (CDK Assets) and that somehow worked.

```
user_data_file_spec = self.node.try_get_context('user-data-file-spec')
user_data_tmp_file_spec = self.node.try_get_context('user-data-tmp-file-spec') 
...
work_dirname = os.path.dirname(__file__)
        cdk_asset = aws_s3_assets.Asset(self, "Asset", 
                                        path=os.path.join(work_dirname, user_data_file_spec))
...
local_path = instance.user_data.add_s3_download_command(bucket=cdk_asset.bucket, 
                                                                bucket_key=cdk_asset.s3_object_key,
                                                                local_file=user_data_tmp_file_spec
                                                                )
```
